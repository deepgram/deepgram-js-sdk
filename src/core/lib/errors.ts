export class DeepgramError extends Error {
  protected __dgError = true;

  constructor(message: string) {
    super(message);
    this.name = "DeepgramError";
  }
}

export function isDeepgramError(error: unknown): error is DeepgramError {
  return typeof error === "object" && error !== null && "__dgError" in error;
}

export class DeepgramApiError extends DeepgramError {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = "DeepgramApiError";
    this.status = status;
  }

  toJSON() {
    return {
      name: this.name,
      message: this.message,
      status: this.status,
    };
  }
}

export class DeepgramUnknownError extends DeepgramError {
  originalError: unknown;

  constructor(message: string, originalError: unknown) {
    super(message);
    this.name = "DeepgramUnknownError";
    this.originalError = originalError;
  }
}

export class DeepgramVersionError extends DeepgramError {
  constructor() {
    super(
      `You are attempting to use an old format for a newer SDK version. Read more here: https://dpgr.am/js-v3`
    );

    this.name = "DeepgramVersionError";
  }
}

/**
 * Enhanced WebSocket error that captures additional debugging information
 * including status codes, request IDs, and response headers when available.
 */
export class DeepgramWebSocketError extends DeepgramError {
  originalEvent?: ErrorEvent | Event;
  statusCode?: number;
  requestId?: string;
  responseHeaders?: Record<string, string>;
  url?: string;
  readyState?: number;

  constructor(
    message: string,
    options: {
      originalEvent?: ErrorEvent | Event;
      statusCode?: number;
      requestId?: string;
      responseHeaders?: Record<string, string>;
      url?: string;
      readyState?: number;
    } = {}
  ) {
    super(message);
    this.name = "DeepgramWebSocketError";
    this.originalEvent = options.originalEvent;
    this.statusCode = options.statusCode;
    this.requestId = options.requestId;
    this.responseHeaders = options.responseHeaders;
    this.url = options.url;
    this.readyState = options.readyState;
  }

  toJSON() {
    return {
      name: this.name,
      message: this.message,
      statusCode: this.statusCode,
      requestId: this.requestId,
      responseHeaders: this.responseHeaders,
      url: this.url,
      readyState: this.readyState,
      originalEvent: this.originalEvent
        ? {
            type: this.originalEvent.type,
            timeStamp: this.originalEvent.timeStamp,
          }
        : undefined,
    };
  }
}
