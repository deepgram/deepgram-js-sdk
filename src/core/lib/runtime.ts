export const NODE_VERSION =
  typeof process !== "undefined" && process.versions && process.versions.node
    ? process.versions.node
    : "unknown";

export const BUN_VERSION =
  typeof process !== "undefined" && process.versions && process.versions.bun
    ? process.versions.bun
    : "unknown";

export const BROWSER_AGENT =
  typeof window !== "undefined" && window.navigator && window.navigator.userAgent
    ? window.navigator.userAgent
    : "unknown";

export const isBrowser = () => BROWSER_AGENT !== "unknown";

export const isNode = () => NODE_VERSION !== "unknown";

export const isBun = () => BUN_VERSION !== "unknown";
