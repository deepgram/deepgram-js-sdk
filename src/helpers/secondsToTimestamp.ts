export function secondsToTimestamp(seconds: number): string {
  return new Date(seconds * 1000).toISOString().substr(11, 12);
}
