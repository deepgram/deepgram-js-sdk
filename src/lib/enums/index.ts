export { LiveConnectionState } from "./LiveConnectionState";
export { LiveTranscriptionEvents } from "./LiveTranscriptionEvents";
