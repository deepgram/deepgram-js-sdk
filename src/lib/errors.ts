export class DeepgramError extends Error {
  protected __dgError = true;

  constructor(message: string) {
    super(message);
    this.name = "DeepgramError";
  }
}

export function isDeepgramError(error: unknown): error is DeepgramError {
  return typeof error === "object" && error !== null && "__dgError" in error;
}

export class DeepgramApiError extends DeepgramError {
  status: number;

  constructor(message: string, status: number) {
    super(message);
    this.name = "DeepgramApiError";
    this.status = status;
  }

  toJSON() {
    return {
      name: this.name,
      message: this.message,
      status: this.status,
    };
  }
}

export class DeepgramUnknownError extends DeepgramError {
  originalError: unknown;

  constructor(message: string, originalError: unknown) {
    super(message);
    this.name = "DeepgramUnknownError";
    this.originalError = originalError;
  }
}

export class DeepgramVersionError extends DeepgramError {
  constructor() {
    super(
      `You are attempting to use an old format for a newer SDK version. Read more here: https://dpgr.am/js-v3`
    );

    this.name = "DeepgramVersionError";
  }
}
