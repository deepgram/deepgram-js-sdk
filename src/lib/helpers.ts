import { Headers as CrossFetchHeaders } from "cross-fetch";
import {
  DeepgramClientOptions,
  FileSource,
  PrerecordedSource,
  UrlSource,
  TextSource,
  AnalyzeSource,
} from "./types";
import { Readable } from "stream";
import merge from "deepmerge";

export function stripTrailingSlash(url: string): string {
  return url.replace(/\/$/, "");
}

export const isBrowser = () => typeof window !== "undefined";
export const isServer = () => typeof process !== "undefined";

export function applySettingDefaults(
  options: DeepgramClientOptions,
  defaults: DeepgramClientOptions
): DeepgramClientOptions {
  return merge(defaults, options);
}

export function appendSearchParams(
  searchParams: URLSearchParams,
  options: Record<string, unknown>
): void {
  Object.keys(options).forEach((i) => {
    if (Array.isArray(options[i])) {
      const arrayParams = options[i] as Array<any>;
      arrayParams.forEach((param) => {
        searchParams.append(i, String(param));
      });
    } else {
      searchParams.append(i, String(options[i]));
    }
  });
}

export const resolveHeadersConstructor = () => {
  if (typeof Headers === "undefined") {
    return CrossFetchHeaders;
  }

  return Headers;
};

export const isUrlSource = (
  providedSource: PrerecordedSource | AnalyzeSource
): providedSource is UrlSource => {
  if ((providedSource as UrlSource).url) return true;

  return false;
};

export const isTextSource = (
  providedSource: PrerecordedSource | AnalyzeSource
): providedSource is TextSource => {
  if ((providedSource as TextSource).text) return true;

  return false;
};

export const isFileSource = (providedSource: PrerecordedSource): providedSource is FileSource => {
  if (isReadStreamSource(providedSource) || isBufferSource(providedSource)) return true;

  return false;
};

const isBufferSource = (providedSource: PrerecordedSource): providedSource is Buffer => {
  if (providedSource as Buffer) return true;

  return false;
};

const isReadStreamSource = (providedSource: PrerecordedSource): providedSource is Readable => {
  if (providedSource as Readable) return true;

  return false;
};

export class CallbackUrl extends URL {
  private callbackUrl = true;
}
