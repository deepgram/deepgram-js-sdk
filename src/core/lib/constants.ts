import { convertProtocolToWs } from "./helpers";
import { isBrowser, isBun, isNode, NODE_VERSION, BUN_VERSION, BROWSER_AGENT } from "./runtime";
import { version } from "./version";
import type { DefaultNamespaceOptions, DefaultClientOptions } from "./types";

const getAgent = () => {
  if (isNode()) {
    return `node/${NODE_VERSION}`;
  } else if (isBun()) {
    return `bun/${BUN_VERSION}`;
  } else if (isBrowser()) {
    return `javascript ${BROWSER_AGENT}`;
  } else {
    return `unknown`;
  }
};

export const DEFAULT_HEADERS = {
  "Content-Type": `application/json`,
  "X-Client-Info": `@deepgram/sdk; ${isBrowser() ? "browser" : "server"}; v${version}`,
  "User-Agent": `@deepgram/sdk/${version} ${getAgent()}`,
};

export const DEFAULT_URL = "https://api.deepgram.com";
export const DEFAULT_AGENT_URL = "wss://agent.deepgram.com";

export const DEFAULT_GLOBAL_OPTIONS: Partial<DefaultNamespaceOptions> = {
  fetch: { options: { url: DEFAULT_URL, headers: DEFAULT_HEADERS } },
  websocket: {
    options: { url: convertProtocolToWs(DEFAULT_URL), _nodeOnlyHeaders: DEFAULT_HEADERS },
  },
};

export const DEFAULT_AGENT_OPTIONS: Partial<DefaultNamespaceOptions> = {
  fetch: { options: { url: DEFAULT_URL, headers: DEFAULT_HEADERS } },
  websocket: {
    options: { url: DEFAULT_AGENT_URL, _nodeOnlyHeaders: DEFAULT_HEADERS },
  },
};

export const DEFAULT_OPTIONS: DefaultClientOptions = {
  global: DEFAULT_GLOBAL_OPTIONS,
  agent: DEFAULT_AGENT_OPTIONS,
};

export enum SOCKET_STATES {
  connecting = 0,
  open = 1,
  closing = 2,
  closed = 3,
}

export enum CONNECTION_STATE {
  Connecting = "connecting",
  Open = "open",
  Closing = "closing",
  Closed = "closed",
}
