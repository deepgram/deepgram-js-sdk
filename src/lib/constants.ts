import { isBrowser } from "./helpers";
import { FetchOptions } from "./types/Fetch";
import { version } from "./version";

export const DEFAULT_HEADERS = {
  "Content-Type": `application/json`,
  "X-Client-Info": `@deepgram/sdk; ${isBrowser() ? "browser" : "server"}; v${version}`,
  "User-Agent": `@deepgram/sdk/${version}`,
};

export const DEFAULT_URL = "https://api.deepgram.com";

export const DEFAULT_GLOBAL_OPTIONS = {
  url: DEFAULT_URL,
};

export const DEFAULT_FETCH_OPTIONS: FetchOptions = {
  headers: DEFAULT_HEADERS,
};

export const DEFAULT_OPTIONS = {
  global: DEFAULT_GLOBAL_OPTIONS,
  fetch: DEFAULT_FETCH_OPTIONS,
};
