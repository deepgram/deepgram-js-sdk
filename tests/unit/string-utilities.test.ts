import { stripTrailingSlash, convertProtocolToWs } from "../../src/lib/helpers";

describe("Unit Tests - String and URL Utilities", () => {
  describe("stripTrailingSlash", () => {
    it("should remove trailing slash from URLs", () => {
      expect(stripTrailingSlash("https://api.deepgram.com/")).toBe("https://api.deepgram.com");
      expect(stripTrailingSlash("https://api.deepgram.com/v1/")).toBe(
        "https://api.deepgram.com/v1"
      );
      expect(stripTrailingSlash("path/")).toBe("path");
      expect(stripTrailingSlash("/")).toBe("");
    });

    it("should not modify URLs without trailing slash", () => {
      expect(stripTrailingSlash("https://api.deepgram.com")).toBe("https://api.deepgram.com");
      expect(stripTrailingSlash("https://api.deepgram.com/v1")).toBe("https://api.deepgram.com/v1");
      expect(stripTrailingSlash("path")).toBe("path");
    });

    it("should handle empty string", () => {
      expect(stripTrailingSlash("")).toBe("");
    });

    it("should only remove one trailing slash", () => {
      expect(stripTrailingSlash("https://api.deepgram.com//")).toBe("https://api.deepgram.com/");
    });
  });

  describe("convertProtocolToWs", () => {
    it("should convert HTTP to WebSocket protocols", () => {
      expect(convertProtocolToWs("https://api.deepgram.com")).toBe("wss://api.deepgram.com");
      expect(convertProtocolToWs("http://api.deepgram.com")).toBe("ws://api.deepgram.com");
    });

    it("should handle uppercase protocols", () => {
      expect(convertProtocolToWs("HTTPS://API.DEEPGRAM.COM")).toBe("wss://api.deepgram.com");
      expect(convertProtocolToWs("HTTP://API.DEEPGRAM.COM")).toBe("ws://api.deepgram.com");
    });

    it("should not modify WebSocket protocols", () => {
      expect(convertProtocolToWs("wss://already-websocket.com")).toBe(
        "wss://already-websocket.com"
      );
      expect(convertProtocolToWs("ws://already-websocket.com")).toBe("ws://already-websocket.com");
    });

    it("should handle mixed case protocols", () => {
      expect(convertProtocolToWs("Http://example.com")).toBe("ws://example.com");
      expect(convertProtocolToWs("Https://example.com")).toBe("wss://example.com");
    });

    it("should handle protocols without authority", () => {
      expect(convertProtocolToWs("http:example")).toBe("ws:example");
      expect(convertProtocolToWs("https:example")).toBe("wss:example");
    });
  });
});
