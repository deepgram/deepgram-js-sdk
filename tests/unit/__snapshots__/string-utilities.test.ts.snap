// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit Tests - String and URL Utilities convertProtocolToWs should convert HTTP protocols to WebSocket protocols 1`] = `
[
  {
    "input": "https://api.deepgram.com",
    "output": "wss://api.deepgram.com",
  },
  {
    "input": "http://api.deepgram.com",
    "output": "ws://api.deepgram.com",
  },
  {
    "input": "HTTPS://API.DEEPGRAM.COM",
    "output": "wss://api.deepgram.com",
  },
  {
    "input": "HTTP://API.DEEPGRAM.COM",
    "output": "ws://api.deepgram.com",
  },
  {
    "input": "wss://already-websocket.com",
    "output": "wss://already-websocket.com",
  },
  {
    "input": "ws://already-websocket.com",
    "output": "ws://already-websocket.com",
  },
]
`;

exports[`Unit Tests - String and URL Utilities stripTrailingSlash should handle various URL formats 1`] = `
[
  {
    "input": "https://api.deepgram.com/",
    "output": "https://api.deepgram.com",
  },
  {
    "input": "https://api.deepgram.com",
    "output": "https://api.deepgram.com",
  },
  {
    "input": "https://api.deepgram.com/v1/",
    "output": "https://api.deepgram.com/v1",
  },
  {
    "input": "https://api.deepgram.com/v1",
    "output": "https://api.deepgram.com/v1",
  },
  {
    "input": "/",
    "output": "",
  },
  {
    "input": "",
    "output": "",
  },
  {
    "input": "path/",
    "output": "path",
  },
  {
    "input": "path",
    "output": "path",
  },
]
`;
